scipy/cluster/_structures.pxi

```diff
@@ -0,0 +1,89 @@
+import numpy as np
+
+
+ctypedef struct Pair:
+    int key
+    double value
+
+
+cdef class Heap:
+    """Binary heap.
+
+    Heap stores values and keys. Values are passed explicitly, whereas keys
+    are assigned implicitly to natural numbers (from 0 to n - 1).
+
+    The supported operations (all have O(log n) time complexity):
+
+        * Return the current minimum value and the corresponding key.
+        * Remove the current minimum value.
+        * Change the value of the given key. Note that the key must be still
+          in the heap.
+    """
+    cdef int[:] index_by_key
+    cdef int[:] key_by_index
+    cdef double[:] values
+    cdef int size
+
+    def __init__(self, double[:] values):
+        self.size = values.shape[0]
+        self.index_by_key = np.arange(self.size, dtype=np.intc)
+        self.key_by_index = np.arange(self.size, dtype=np.intc)
+        self.values = values.copy()
+        cdef int i
+        for i in reversed(range(self.size / 2)):
```

Comment:
Yes, it's a C division. I rely on the directive in `_hierarchy.pyx`, now I realize it's better to include it here as well for safety.

This loop is the known algorithm of building a heap from unordered data in a linear time. I can't quickly explain why it works, the article in Wikipedia on heap does that for example.